lib/model/message.dart

```diff
@@ -1,21 +1,332 @@
+  /// the outbox message's timer gets removed and cancelled.
+  final Map<int, Timer> _outboxMessageWaitPeriodTimers = {};
+
+  /// A fresh ID to use for [OutboxMessage.localMessageId],
+  /// unique within this instance.
+  int _nextLocalMessageId = 0;
+
+  /// As in [MessageStoreImpl._messageListViews].
+  Set<MessageListView> get _messageListViews;
+
+  /// As in [MessageStoreImpl._disposed].
+  bool get _disposed;
+
+  /// Update the state of the [OutboxMessage] with the given [localMessageId],
+  /// and notify listeners if necessary.
+  ///
+  /// The outbox message with [localMessageId] must exist.
+  void _updateOutboxMessage(int localMessageId, {
+    required OutboxMessageState newState,
+  }) {
+    assert(!_disposed);
+    final outboxMessage = outboxMessages[localMessageId];
+    if (outboxMessage == null) {
+      throw StateError(
+        'Removing unknown outbox message with localMessageId: $localMessageId');
+    }
+    final oldState = outboxMessage.state;
+    // See [OutboxMessageState] for valid state transitions.
+    final isStateTransitionValid = newState != oldState && switch (newState) {
+      OutboxMessageState.hidden => false,
+      OutboxMessageState.waiting =>
+        oldState == OutboxMessageState.hidden,
+      OutboxMessageState.waitPeriodExpired =>
+        oldState == OutboxMessageState.waiting,
+      OutboxMessageState.failed =>
+        oldState == OutboxMessageState.hidden
+        || oldState == OutboxMessageState.waiting
+        || oldState == OutboxMessageState.waitPeriodExpired,
+    };
+    if (!isStateTransitionValid) {
+      throw StateError('Unexpected state transition: $oldState -> $newState');
+    }
+
+    outboxMessage._state = newState;
+    for (final view in _messageListViews) {
+      if (oldState == OutboxMessageState.hidden) {
+        view.addOutboxMessage(outboxMessage);
+      } else {
+        view.notifyListenersIfOutboxMessagePresent(localMessageId);
+      }
+    }
+  }
+
+  /// Send a message and create an entry of [OutboxMessage].
+  Future<void> _outboxSendMessage({
+    required MessageDestination destination,
+    required String content,
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    assert(!_disposed);
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final now = ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000;
+    _outboxMessages[localMessageId] = switch (destination) {
+      StreamDestination(:final streamId, :final topic) => StreamOutboxMessage(
+        localMessageId: localMessageId,
+        selfUserId: selfUserId,
+        timestamp: now,
+        conversation: StreamConversation(
+          streamId,
+          topic.processLikeServer(
+            // Processing this just once on creating the outbox message
+            // allows an uncommon bug, because either of these values can change.
+            // During the outbox message's life, a topic processed from
```

Comment:
Let's move this `topic.processLikeServer` call out to a little private helper method. That will keep the flow of the rest of this method's logic a bit easier to read.